core/src/apps/zcash/sig_hasher.py

```diff
@@ -0,0 +1,187 @@
+        self.prevouts =      HashWriter(blake2b(outlen=32, personal=b'ZTxIdPrevoutHash'))
+        self.amounts  =      HashWriter(blake2b(outlen=32, personal=b'ZTxTrAmountsHash'))
+        self.scriptpubkeys = HashWriter(blake2b(outlen=32, personal=b'ZTxTrScriptsHash'))
+        self.sequence =      HashWriter(blake2b(outlen=32, personal=b'ZTxIdSequencHash'))
+        self.outputs  =      HashWriter(blake2b(outlen=32, personal=b'ZTxIdOutputsHash'))
+        self.empty = True  # inputs_amount + outputs_amount == 0
+
+    def add_input(self, txi: TxInput, script_pubkey: bytes) -> None:
+        self.empty = False
+        write_prevout(self.prevouts, txi)                        # (see S.2b prevouts_sig_digest)
+        write_uint64(self.amounts, txi.amount)                   # (see S.2c amounts_sig_digest) 
+        write_bytes_prefixed(self.scriptpubkeys, script_pubkey)  # (see S.2d scriptpubkeys_sig_digest)
+        write_uint32(self.sequence, txi.sequence)                # (see S.2e sequence_sig_digest)
+
+    def add_output(self, txo: TxOutput, script_pubkey: bytes) -> None:
+        self.empty = False
+        write_tx_output(self.outputs, txo, script_pubkey)        # (see S.2f outputs_sig_digest)
+
+    def digest(self):
+        """Returns `T.2: transparent_digest` field for txid computation."""
+        h = HashWriter(blake2b(outlen=32, personal=b'ZTxIdTranspaHash'))
+
+        if not self.empty:
+            write_hash(h, self.prevouts.get_digest())  # T.2a: prevouts_digest (32-byte hash)
+            write_hash(h, self.sequence.get_digest())  # T.2b: sequence_digest (32-byte hash)
+            write_hash(h, self.outputs.get_digest())   # T.2c: outputs_digest  (32-byte hash)
+
+        return h.get_digest()
+
+    def sig_digest(self, txin_sig_digest: bytes | None):
+        """Returns `S.2: transparent_sig_digest` field for signature digest computation."""
+        if self.empty:
+            assert txin_sig_digest is None
+            return self.digest()
+
+        if txin_sig_digest is None:
+            txin_sig_digest = blake2b(outlen=32, personal=b'Zcash___TxInHash').digest()
+
+        h = HashWriter(blake2b(outlen=32, personal=b'ZTxIdTranspaHash'))
+
+        # only SIGHASH_ALL is supported in Trezor
+        write_uint8(h, SigHashType.SIGHASH_ALL)         # S.2a: hash_type                (1 byte)
+        write_hash(h, self.prevouts.get_digest())       # S.2b: prevouts_sig_digest      (32-byte hash)
+        write_hash(h, self.amounts.get_digest())        # S.2c: amounts_sig_digest       (32-byte hash)
+        write_hash(h, self.scriptpubkeys.get_digest())  # S.2d: scriptpubkeys_sig_digest (32-byte hash)
+        write_hash(h, self.sequence.get_digest())       # S.2e: sequence_sig_digest      (32-byte hash)
+        write_hash(h, self.outputs.get_digest())        # S.2f: outputs_sig_digest       (32-byte hash)
+        write_hash(h, txin_sig_digest)                  # S.2g: txin_sig_digest          (32-byte hash)
```

Comment:
I think the comments should go on a separate line above the code, because the code formatter will break them (not only here). Run `make style` in the root directory to format the code.

Reply:
Yay , I didn't know we have a style formater. Ok, I will be using it.

Reply:
More information/guidelines about contributing to Trezor: https://github.com/trezor/trezor-firmware/blob/master/docs/misc/contributing.md.